kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/KonanBCEForLoopBodyTransformer.kt

```diff
@@ -0,0 +1,244 @@
+                        } else if (loopHeader.headerInfo.last.compareIntegerNumericConst { it >= valueToCompare }) {
+                            boundsAreSafe = true
+                        }
+                        if (!boundsAreSafe)
+                            return analysisResult
+                        when (loopHeader.headerInfo.first) {
+                            is IrCall -> {
+                                val functionCall = (loopHeader.headerInfo.first as IrCall)
+                                analysisResult = lessThanSize(functionCall)
+                            }
+                            is IrGetValue -> {
+                                (((loopHeader.headerInfo.first as IrGetValue).symbol.owner as? IrVariable)?.initializer as? IrCall)?.let {
+                                    analysisResult = lessThanSize(it)
+                                }
+                            }
+                        }
+                    }
+                    ProgressionDirection.UNKNOWN ->
+                        // Case of progression - for (i in 0 until array.size step n)
+                        if (loopHeader.headerInfo.first.isProgressionPropertyGetter("first") &&
+                                loopHeader.headerInfo.last.isProgressionPropertyGetter("last")) {
+                            val firstReceiver = (loopHeader.headerInfo.first as IrCall).dispatchReceiver as? IrGetValue
+                            val lastReceiver = (loopHeader.headerInfo.last as IrCall).dispatchReceiver as? IrGetValue
+                            if (firstReceiver?.symbol?.owner == lastReceiver?.symbol?.owner) {
+                                val untilFunction =
+                                        ((firstReceiver?.symbol?.owner as? IrVariable)?.initializer as? IrCall)?.extensionReceiver as? IrCall
+                                if (untilFunction?.symbol?.owner?.name?.asString() == "until" && untilFunction.extensionReceiver?.compareIntegerNumericConst { it >= 0 } == true) {
+                                    val last = untilFunction.getValueArgument(0)!!
+                                    if (last is IrCall) {
+                                        analysisResult = lessThanSize(last)
+                                        // `isLastInclusive` for current case is set to true.
+                                        // This case isn't fully optimized in ForLoopsLowering.
+                                        if (last.isGetSizeCall()) {
+                                            analysisResult = BoundsCheckAnalysisResult(true, (last.dispatchReceiver as? IrGetValue)?.symbol)
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                }
+
+            is WithIndexLoopHeader ->
+                when(loopHeader.nestedLoopHeader) {
+                    is IndexedGetLoopHeader -> {
```

Comment:
`IndexedGetLoopHeader` is also used for `CharSequence` and `String`. Did you intend to also support those cases? e.g.,
```kt
for ((i, c) in "foo".withIndex()) { ... }
```

Reply:
These cases should be skipped because there is a check of  dispatch receiver type. Or may be I didn't get right the problem case?